compiler/haskell-ide-core/src/Development/IDE/Types/Diagnostics.hs

```diff
@@ -184,3 +199,170 @@ prettyFileDiagnostics (filePath, diags) =
 getDiagnosticsFromStore :: StoreItem -> [Diagnostic]
 getDiagnosticsFromStore (StoreItem _ diags) =
     toList =<< Map.elems diags
+
+-- | A wrapper around the lsp diagnostics store, the constraint describes how compilation steps
+--   are represented
+newtype Diagnostics stage = Diagnostics {getStore :: DiagnosticStore}
+    deriving Show
+
+instance Semigroup (Diagnostics stage) where
+    (Diagnostics a) <> (Diagnostics b) =
+        Diagnostics $
+        Map.unionWith (curry combineStores)
+        a b where
+        combineStores = \case
+            (StoreItem Nothing x, StoreItem Nothing y) ->
+                StoreItem Nothing $ Map.unionWith SL.union x y
+            _ -> noVersions
+
+instance Monoid (Diagnostics k) where
+    mempty = emptyDiagnostics
+
+prettyDiagnostics :: Diagnostics stage -> Doc SyntaxClass
+prettyDiagnostics ds =
+    slabel_ "Compiler errors in" $ vcat $ concatMap fileErrors storeContents where
+
+    fileErrors :: (FilePath, [(T.Text, [LSP.Diagnostic])]) -> [Doc SyntaxClass]
+    fileErrors (filePath, stages) =
+        [ slabel_ "File:" $ pretty filePath
+        , slabel_ "Errors:" $ vcat $ map prettyStage stages
+        ]
+
+    prettyStage :: (T.Text, [LSP.Diagnostic]) -> Doc SyntaxClass
+    prettyStage (stage,diags) =
+        slabel_ ("Stage: "<>T.unpack stage) $ vcat $ concatMap diagSC diags
+
+    storeContents ::
+        [(FilePath, [(T.Text, [LSP.Diagnostic])])]
+        -- ^ Source File, Stage Source, Diags
+    storeContents =
+        map (\(uri, StoreItem _ si) ->
+                 (fromMaybe dontKnow $ uriToFilePath uri, getDiags si))
+            $ Map.assocs
+            $ getStore
+            $ removeEmptyStages ds
+
+    dontKnow :: IsString s => s
```

Comment:
What is the difference between `dontKnow` and `noFilePath`?

Reply:
Dontknow